pkg/credentialprovider/plugin/plugin.go

```diff
@@ -0,0 +1,215 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package plugin
+
+import (
+	"bytes"
+	"fmt"
+	"os/exec"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/klog/v2"
+	"k8s.io/kubernetes/pkg/credentialprovider"
+	"k8s.io/kubernetes/pkg/credentialprovider/apis/criauth"
+	"k8s.io/kubernetes/pkg/credentialprovider/apis/criauth/install"
+	"k8s.io/kubernetes/pkg/credentialprovider/apis/criauth/v1alpha1"
+)
+
+var (
+	scheme = runtime.NewScheme()
+	codecs = serializer.NewCodecFactory(scheme)
+)
+
+func init() {
+	install.Install(scheme)
+}
+
+// RegisterCRIAuthPlugins is called from kubelet to register external CRI auth
+// plugins according to the CRI auth plugin config file.
+func RegisterCRIAuthPlugins(pluginConfigFile, pluginBinDir string) error {
+	criAuthPluginConfig, err := readCRIAuthPluginConfigFile(pluginConfigFile)
+	if err != nil {
+		return err
+	}
+
+	// For a given plugin configuration, each image match reference is registered
+	// as separate credential provider to the keyring.
+	for _, plugin := range criAuthPluginConfig.Plugins {
+		for _, imageMatchRef := range plugin.ImageMatchRefs {
+			cachingProvider := &credentialprovider.CachingDockerConfigProvider{
+				Provider: &pluginProvider{
+					matchURL: imageMatchRef.MatchURL,
+					plugin: &execPlugin{
+						name:         plugin.Name,
+						pluginBinDir: pluginBinDir,
+						args:         convertArgs(plugin.ExtraArgs),
+					},
+				},
+				Lifetime: imageMatchRef.CacheDuration.Duration,
+			}
+
+			// The registered provider name is the plugin name + the match URL of the image reference since
+			// each image match reference in a given plugin is registered as a separate credential provider
+			// and each registered name must be unique.
+			providerName := plugin.Name + "-" + imageMatchRef.MatchURL
+			credentialprovider.RegisterCredentialProvider(providerName, cachingProvider)
```

Comment:
this is a map, not a list of providers. So my understanding is that the ordering of plugins is getting lost here. It may be an issue

Reply:
Hmm, good call. So the ordering is only guarenteed between external plugins and the built-in ones here: https://github.com/kubernetes/kubernetes/pull/94196/files#diff-8bcfde5fa58ec8d8ae4b306dd85768a2R53-R84

Order only amongst the external plugins and the builtin plugins is not guarenteeed.

Reply:
this is definitely not a default expectation from configurations with the matching urls like this. Typically ordering is assumed. I think this either needs to be very clearly outlined in the configuration file that the ordering is not guaranteed or fixed.

Reply:
I think this is fixed in a later commit now -- auth plugins will always run before built-in plugins. Plugin providers and built-in providers are then ordered by string, which is the current behavior today for built-in providers:

https://github.com/kubernetes/kubernetes/442a69c3bdf6fe8e525b05887e57d89db1e2f3a5/master/pkg/credentialprovider/plugins.go#L54-L68